compaction/compaction.cc

```diff
@@ -746,6 +746,45 @@ class compaction {
         return consumer(make_compacting_reader(setup_sstable_reader(), compaction_time, max_purgeable_func(), gc_state));
     }
 
+    // when compacting a run of mutation fragments, the producer reads the
+    // mutations from sstables (or yet another reader). and the consumer
+    // processes the mutations, by either compacting them, or segregates them
+    // into chunks. each chunk of the fragments stream can be either piped to
+    // a sstable writer or yet another consumer. the consumers and writers can
+    // be created on demand.
+    //
+    //   def sstable_reader(sstables):
+    //       for sstable in sstables:
+    //           for frag in read(sstable):
+    //               yield frag
+    //
+    //   def compacting_reader(frag_gen):
+    //       for frag in compact(frag_gen):
+    //           yield frag
+    //
+    //   def interposer_consumer(frag_gen):
+    //       for frag in segregate(frag_gen):
+    //           yield frag
```

Comment:
This doesn't reflect what segregate does. Segregate does multiplexing: splitting a single stream into N (possibly parallel) output streams, based on some criteria.

Reply:
The algorithm is more like the following:
```python
buckets = dict() : Dict[bucket_id, bucket]
for frag in frag_gen:
    bucked_id = classify(frag)
    if not bucket_id in buckets:
        create_new_bucket(bucket_id)
    buckets[bucket].consume(frag)
```

Reply:
Where each `bucket` is usually an end consumer (sstable writer).

Reply:
> The algorithm is more like the following:
> 
> ```python
> buckets = dict() : Dict[bucket_id, bucket]
> for frag in frag_gen:
>     bucked_id = classify(frag)
>     if not bucket_id in buckets:
>         create_new_bucket(bucket_id)
>     buckets[bucket].consume(frag)
> ```

thank you, Botond! just saw your comment after posting my last comment with the revised code. i think `create_new_bucket()` should use the `end_consumer` created by the base class of `compaction` right?

Reply:
Yes.

Reply:
> do you mean that **new** streams can be created in the middle of processing ? and we cannot foresee the number (or the range of "bucket_id" in your code) of streams and assign them to their consumers respectively, even if we just need to pass a generator to each of them?

Correct. We cannot determine the number of streams we will have. They are opened/closed on-demand in the middle of processing the input stream, based on the data at hand.

Reply:
i see. will update the comment to reflect this.